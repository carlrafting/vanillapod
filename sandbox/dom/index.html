<!doctype html>
<html lang="en">

<head>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>Vanillapod DOM Sandbox</title>
</head>

<body>

  <div id="test">
    I will be overriden!
  </div>

  <div id="app" class="flow">
  </div>

  <script src="./main.js" type="module"></script>

</body>

</html>
