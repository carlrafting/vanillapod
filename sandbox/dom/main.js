import {
    createMountable,
    div,
    p as $p,
    h1,
    button as $button,
    render,
} from 'vanillapod/dom';
import { createSignal, createEffect } from 'vanillapod/state';

console.time('dom');

const root = document.querySelector('#app');

const nested = createMountable(
    'div',
    { className: 'hero' },
    createMountable('h1', 'Hi Again!')
);
const p = createMountable('p', 'Hello World!');
const list = createMountable(
    'ul',
    createMountable('li', 'One'),
    createMountable('li', 'Two'),
    createMountable('li', 'Three')
);
const elWithProps = createMountable(
    'div',
    { className: 'red green yellow', id: 'el-with-props' },
    'An element with props!'
);
const textInput = createMountable('input', {
    id: 'text-input',
    value: 'Write Something!',
    type: 'text',
});
const button = createMountable(
    'button',
    {
        id: 'button',
        type: 'button',
        onClick: (event) => {
            alert('CLICKED!');
            console.log(event);
            event.preventDefault();
        },
    },
    'Save'
);

function MyComponent() {
    return div('This is a component!', {
        id: 'my-component',
        style: 'background-color:yellow;padding:1rem;',
    });
}

function ComponentHasProps({ text }) {
    return div(
        {
            id: 'component-with-props',
            style: 'background-color:red;padding:1rem;',
        },
        h1(text),
        $p('And a paragraph inside!')
    );
}

function ButtonWithSignals() {
    const [text, setText] = createSignal('Update Signal');
    // createEffect(() => alert(text()));
    return $button(
        {
            onClick(e) {
                setText('Hello Signals!');
                e.target.textContent = text(); // quick & dirty solution until i figure out something better
            },
        },
        text()
    );
}

render(
    () => root,
    nested,
    p,
    list,
    elWithProps,
    textInput,
    button,
    div({ id: 'helper' }, 'Helper Method'),
    div({ id: 'foobar' }, '123'),
    div(
        { id: 'multiple-prop-defs' },
        'Merge Multiple Property Definitions',
        {
            style: 'background-color:green;',
        },
        { style: 'padding:.5rem' }
    ),
    MyComponent,
    [ComponentHasProps, { text: 'This component has props!' }],
    ButtonWithSignals,
    createMountable(
        document.getElementById('test'),
        { id: 'test-overridden' },
        'Enhanced'
    )
);

console.timeEnd('dom');
