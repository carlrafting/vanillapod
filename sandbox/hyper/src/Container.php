<?php

declare(strict_types=1);

namespace Vanillapod\Hyper;

class ContainerException extends \ErrorException implements \Psr\Container\ContainerExceptionInterface
{
}

class ContainerNotFound extends ContainerException implements \Psr\Container\NotFoundExceptionInterface
{
}

class Container implements \Psr\Container\ContainerInterface
{
    public function __construct(private array $containers = [])
    {
    }

    public function add(string $id, mixed $entry): mixed
    {
        if (!$this->has($id)) {
            return $this->containers[$id] = $entry;
        }

        throw new ContainerException("There is already a container registered with id: $id");
    }
    public function get(string $id): mixed
    {
        if ($this->has($id)) {
            return $this->containers[$id];
        }

        throw new ContainerNotFound("No container could be found with id: $id");
    }
    public function has(string $id): bool
    {
        if (array_key_exists($id, $this->containers)) {
            return true;
        }

        return false;
    }
}
