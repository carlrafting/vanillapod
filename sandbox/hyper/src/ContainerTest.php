<?php

declare(strict_types=1);

use PHPUnit\Framework\TestCase;
use PHPUnit\Framework\Attributes\CoversClass;

#[CoversClass(\Vanillapod\Hyper\Container::class)]
final class ContainerTest extends TestCase
{
    public function testBasicUseCase()
    {
        $container = new \Vanillapod\Hyper\Container();
        $container->add('foo', function () {
            return "hello foo!";
        });
        $foo = $container->get('foo');
        $this->assertEquals($foo(), "hello foo!");
    }
    public function testErrorCase()
    {
        $container = new \Vanillapod\Hyper\Container();
        $container->add('foo', function () {
            return "hello foo!";
        });
        $this->expectException(\Vanillapod\Hyper\ContainerException::class);
        $container->add('foo', function () {
            return "hello foo!";
        });
    }
}
