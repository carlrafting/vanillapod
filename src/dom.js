import { createMemo } from './memo.js';
import { createError } from './error.js';
import { checkType } from './utils.js';

const log = (...output) => console.log(...output);

console.time('dom/internal');

const eventHandlers = new Set();

// function merge() {}

export function createMountable(element, ...params) {
    const template = document.createElement('template');
    if (typeof element === 'string') {
        element = document.createElement(element);
    }
    if (!element || element === null) {
        element = document.createDocumentFragment();
    }
    if (params.length > 0) {
        let prevProps = null;
        for (const param of params) {
            if (checkType(param) === 'array') {
                const [component, props] = param;
                console.log({ component });
                element.append(component(props));
            }
            if (typeof param === 'function') {
                element.append(param());
            }
            if (typeof param === 'string') {
                const text = document.createTextNode(param);
                element.append(text);
            }
            if (typeof param === 'object' && !(param instanceof Node)) {
                // merge props if there are multiple props definitions
                // does not work if the same key exists on multiple prop definitions
                const props =
                    prevProps !== null
                        ? structuredClone({ ...prevProps, ...param })
                        : { ...param };
                if (prevProps === null) {
                    prevProps = props;
                }
                const indexOfStart = 0;
                const eventKey = 'on';
                for (const [key, value] of Object.entries(props)) {
                    // log({ key, value });
                    // if (key in element) {
                    element[key] = value;
                    // }
                    if (key.indexOf(eventKey) === indexOfStart) {
                        const sliceStart = eventKey.length;
                        const normalizedEvent = key
                            .slice(sliceStart)
                            .toLowerCase();
                        eventHandlers.add({
                            element,
                            event: [normalizedEvent, value],
                        });
                        // continue;
                        /* console.log({
                            normalizedEvent,
                            indexOfStart,
                            sliceStart,
                        }); */
                    }
                }
            }
            // log('bottom', { param });
            if (param instanceof Node) {
                try {
                    element.append(param);
                } catch (err) {
                    console.error(err);
                }
            }
        }
    }
    template.content.append(element);
    // return template.content.cloneNode(true);
    return template.content; // this makes the event handling work since we do not loose the reference to the element.
}

const dom = new Map();

/**
 * elements
 * @url https://developer.mozilla.org/en-US/docs/Web/HTML/Element#interactive_elements
 */
const elements = new Set([
    // 1. Content Sectioning
    'address',
    'article',
    'aside',
    'footer',
    'header',
    'main',
    'nav',
    'section',
    'h1',
    'h2',
    'h3',
    'h4',
    'h5',
    'h6',
    // 2. Text Content
    'div',
    'p',
    'blockquote',
    'ul',
    'ol',
    'li',
    'dl',
    'dt',
    'dd',
    'figure',
    'figcaption',
    'hr',
    'menu',
    'pre',
    // 3. Inline Text Semantics
    'span',
    // 4. Image And Multimedia
    // 5. Embedded content
    // 6. SVG
    // 7. Demarcating Edits
    // 8. Table Content
    // 9. Forms
    'form',
    'fieldset',
    'label',
    'input',
    'textarea',
    'button',
    'select',
    'option',
    // 10. Interactive Elements
    // 11. Web Components
]);

function cleanupElements() {
    if (dom.size > 0) {
        for (const el of elements) {
            elements.delete(el);
        }
        elements.clear();
        // log('cleanupElements', elements);
    }
    createError(`dom: can't cleanup elements Set: ${elements.size > 0}`);
}

function createDOMMethods(done = null) {
    for (const el of elements) {
        dom.set(el, (...params) => {
            const memo = createMemo((...params) =>
                createMountable(el, ...params)
            );
            // return memo(...params); // using memo  creates issues when creating multiple elements with data from an array
            return createMountable(el, ...params);
        });
    }
    if (done && typeof done === 'function') done();
}

createDOMMethods(() => cleanupElements());

// log({ dom });

// log(dom.get('div')('Hello There'));

function createRoot(root = null) {
    if (root instanceof Node) {
        for (const { element, event } of eventHandlers) {
            const [name, fn] = event;
            // log(element, event);
            root.addEventListener(name, (e) => {
                // log({ e, target: e.target, element });
                if (e.target === element) {
                    // log('match!');
                    fn(e);
                }
            });
        }
        return eventHandlers.clear();
    }
    createError('dom: Expected root to be an instance of NodePrototype');
}

export function render(root, ...mountables) {
    if (typeof root === 'function') {
        root = root();
    }
    const results = [];
    const loop = (done = null) => {
        for (let mountable of mountables) {
            if (checkType(mountable) === 'function') {
                results.push(mountable());
                continue;
                // results.push(component);
            }
            if (checkType(mountable) === 'array') {
                const [component, props] = mountable;
                results.push(component(props));
                continue;
            }
            if (mountable || mountable !== null) {
                results.push(mountable);
            }
        }
        // console.log([...mountables]);
        if (done && typeof done === 'function') done();
    };
    loop(function done() {
        root.append(...results);
        results.length = 0; // clear results array since all items are mounted and we no longer have any use for it
        // console.log({ root });
        createRoot(root);
    });
    return function destroy(position = null) {
        if (typeof position !== 'number')
            createError('dom: exprected position to be of value Number');
        const rootChildren = root.querySelectorAll('*');
        if (position === null) {
            // console.log({ root });
            for (const child of rootChildren) {
                child.remove();
            }
            return;
        }
        for (let index = 0; index < rootChildren.length; index++) {
            if (index === position) {
                const mountable = rootChildren[index];
                mountable.remove();
            }
        }
    };
}

function el(el) {
    return (props, ...params) => dom.get(el)(props, ...params);
}

export const div = el('div');
export const p = el('p');
export const h1 = el('h1');
export const h2 = el('h2');
export const h3 = el('h3');
export const h4 = el('h4');
export const h5 = el('h5');
export const h6 = el('h6');
export const article = el('article');
export const aside = el('aside');
export const footer = el('footer');
export const header = el('header');
export const main = el('main');
export const nav = el('nav');
export const section = el('section');
export const ul = el('ul');
export const ol = el('ol');
export const li = el('li');
export const dl = el('dl');
export const dt = el('dt');
export const dd = el('dd');
export const form = el('form');
export const fieldset = el('fieldset');
export const label = el('label');
export const input = el('input');
export const textarea = el('textarea');
export const button = el('button');
export const select = el('select');
export const option = el('option');
export const span = el('span');
export const br = () => document.createElement('br');

console.timeEnd('dom/internal');
