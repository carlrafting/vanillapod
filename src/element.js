import { createError } from './error.js';
import debug from './debug.js';
import setElementProperties from './properties.js';
import setElementAttributes from './attributes.js';
import setElementTextContent from './text.js';
import setElementEventHandlers from './events.js';

export const validProps = {
    element: null,
    el: null,
    data: {},
    classList: [],
    classNames: [],
    attributes: {},
    attrs: {},
    properties: {},
    props: {},
    children: [],
    text: null,
    events: {},
    on: {},
    hooks: {},
};

export function validateProps(props) {
    for (const key in props) {
        if (!Object.hasOwn(validProps, key)) {
            return createError(
                `${key} is not a valid property name. Try one of the following instead: ${Object.keys(
                    validProps
                )}`
            );
        }
    }

    return true;
}

export function createComponentInstance(
    vanillapodComponent,
    vanillapodComponentProps
) {
    if (typeof vanillapodComponent === 'function') {
        debug() && console.log(`Registering ${vanillapodComponent}...`);

        let props = vanillapodComponentProps
            ? vanillapodComponent(vanillapodComponentProps)
            : vanillapodComponent();

        if (typeof props !== 'object' || props === undefined) {
            return createError(`${vanillapodComponent} must return an object`);
        }

        if (validateProps(props)) {
            return {
                vanillapodComponent,
                ...props,
            };
        }
    }

    return createError(`${vanillapodComponent} is not a function`);
}

export function createElement(element, ...args /* props = {} */) {
    let props = {};

    for (const arg of args) {
        if (Array.isArray(arg)) {
            props.children = arg;
            break;
        }
        if (typeof arg === 'function') {
            props = typeof arg() === 'object' && arg();
            break;
        }
        props = arg;
    }

    // console.log(props);

    if (!element) {
        return createError(`element is ${element}`);
    }

    if (typeof element === 'string') {
        // (debug() && console.log(`Creating ${element} for ${props.vanillapodComponent}`));
        debug() && console.log('Creating element...');
        element = document.createElement(element);
        debug() && console.log('Created element: ', element);
    }

    // set element properties
    setElementProperties(element, props);

    // set attributes on elements
    setElementAttributes(element, props);

    // set textContent for element
    setElementTextContent(element, props);

    // register DOM event handlers
    setElementEventHandlers(element, props);

    debug() &&
        console.log(
            'createElement: props',
            props,
            'createElement: element',
            element
        );

    return {
        ...props,
        element,
        el: element,
    };
}

// aliases for createElement
export const h = createElement;
export const e = createElement;
export const html = createElement;
