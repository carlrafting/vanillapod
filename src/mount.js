import { registerHooks, triggerHook } from './hooks.js';
import { createError } from './error.js';
import { checkType } from './utils.js';
import { createElement, createComponentInstance } from './element.js';
import debug from './debug.js';

export default function mount(root, ...args) {
    let index = 0;
    let vanillapodComponent;
    let props = {};
    const argsLength = args.length;

    for (; index < argsLength; index += 1) {
        const arg = args[index];

        [vanillapodComponent, props] = init(arg, vanillapodComponent, props);

        if (!vanillapodComponent) {
            return createError('Could not determine component format');
        }

        const instance = createComponentInstance(vanillapodComponent, props);

        // console.log({ instance });

        if (!instance) {
            return createError(
                `Could not create instance for ${vanillapodComponent}`
            );
        }

        let { el, element } = instance;

        if (!el) {
            el = element;
            element = null;
            // delete instance.element;
        }

        // if el is a string, use implicit render approach
        if (checkType(el) === 'string') {
            const props = createElement(el, instance);
            el = props.el || props.element;
        }

        registerHooks(el, instance);

        const hooks = el._vanillapod_hooks;

        // trigger element before hook
        if (hooks && hooks.before) {
            triggerHook(el, 'before');
        }

        if (instance.children && instance.children.length > 0) {
            mount(el, ...instance.children);
        }

        // mount element to root
        if (root) {
            root.insertBefore(el, null);
        }

        // mount element to document body
        if (!root && document.body) {
            document.body.insertBefore(el, document.body.lastChild);
        }

        // trigger element mount hook
        if (hooks && hooks.mount) {
            triggerHook(el, 'mount');
        }
    }
}

export function init(arg, vanillapodComponent, props) {
    // console.log({arg, vanillapodComponent, props});
    checkType(arg) === 'array' && ([vanillapodComponent, props] = arg);
    checkType(arg) === 'function' && (vanillapodComponent = arg);
    checkType(arg) === 'asyncfunction' && (vanillapodComponent = arg);
    checkType(arg) === 'object' && (vanillapodComponent = () => arg);
    arg instanceof Node && (vanillapodComponent = () => ({ el: arg }));
    return [vanillapodComponent, props];
}
