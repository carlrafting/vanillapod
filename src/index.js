import mount from './mount';
import debug from './debug';
import setElementProperties from './properties';
import setElementAttributes from './attributes';
import setElementTextContent from './text';
import setElementEventHandlers from './events';
import createDocumentFragment from './fragment';

export { memo, createMemo, makeMemo, useMemo } from './memo';
export {
    store,
    createStore,
    makeStore,
    useStore,
    signal,
    createSignal,
    useSignal,
    makeSignal,
} from './state';
export { createError, logErrors } from './error';
export { createComponentInstance, createElement, html, h, e } from './element';
export {
    mount,
    debug,
    setElementProperties,
    setElementAttributes,
    setElementTextContent,
    setElementEventHandlers,
    createDocumentFragment,
};
export { registerHooks, registerHook, triggerHook } from './hooks';
