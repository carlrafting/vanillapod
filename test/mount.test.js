import { test, expect, /* beforeEach, */ } from 'vitest';
import { getByText } from '@testing-library/dom';
import mount from '../src/mount';
// import { debug } from '../src';
import { createElement } from '../src/element';

// beforeEach(() => (document.body.innerHTML = ''));

test('should mount one component correctly', () => {
    function component() {
        return {
            element: 'div',
            text: 'Foo Bar',
        };
    }

    mount(null, component);

    // TODO: look into a more reliable way to test this?
    expect(document.body.firstChild).toMatchSnapshot();
    expect(document.body.firstChild.tagName).toBe('DIV');
    expect(document.body.firstChild).toHaveTextContent('Foo Bar');
});

test('should mount multiple components correctly', () => {
    // debug(true);

    function first() {
        return {
            el: 'div',
            props: {
                id: 'first',
            },
        };
    }

    function second() {
        return {
            el: 'div',
            props: {
                id: 'second',
            },
        };
    }

    mount(document.body, first, second);

    const firstElement = document.getElementById('first');
    const secondElement = document.getElementById('second');

    expect(document.body).toContainElement(firstElement);
    expect(document.body).toContainElement(secondElement);
});

test('should be able to mount explicit component correctly', () => {
    const text = 'Foo Bar Hello World!';

    function explicit() {
        const props = createElement('h1', {
            element: 'h1',
            text,
        });

        return {
            ...props,
        };
    }

    mount(document.body, explicit);

    const el = getByText(document.body, text);

    expect(document.body).toContainElement(el);
    expect(el).toHaveTextContent(text);
    expect(el.nodeName).toEqual('H1');
});

test('should be able to pass props to component when mounting', () => {
    function h1({ message }) {
        return {
            el: 'h1',
            text: message,
        };
    }

    mount(document.body, [h1, { message: 'Hello World!' }]);

    const el = getByText(document.body, 'Hello World!');
    expect(el).toHaveTextContent('Hello World!');
});

test('shoud be able to mount objects matching vanillapod props', () => {
    // console.log(document.body.children);
    const component = { el: 'h1', text: 'Lorem Ipsum!' };
    mount(document.body, component);
    expect(document.body.children).toMatchSnapshot();
    expect(getByText(document.body, component.text)).toHaveTextContent(
        component.text
    );
});

test('shoud be able to mount any kind of node', () => {
    const comment = document.createComment('comment!');
    const fragment = document.createDocumentFragment();
    const text = document.createTextNode('text!');
    mount(document.body, comment, fragment, text);
    expect(document.body).toMatchSnapshot();
});
