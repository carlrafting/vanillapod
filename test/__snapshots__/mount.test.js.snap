// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`shoud be able to mount any kind of node 1`] = `
<body>
  <div>
    Foo Bar
  </div>
  <div
    id="first"
  />
  <div
    id="second"
  />
  <h1>
    Foo Bar Hello World!
  </h1>
  <h1>
    Hello World!
  </h1>
  <h1>
    Lorem Ipsum!
  </h1>
  <!--comment!-->
  text!
</body>
`;

exports[`shoud be able to mount objects matching vanillapod props 1`] = `
[
  <div>
    Foo Bar
  </div>,
  <div
    id="first"
  />,
  <div
    id="second"
  />,
  <h1>
    Foo Bar Hello World!
  </h1>,
  <h1>
    Hello World!
  </h1>,
  <h1>
    Lorem Ipsum!
  </h1>,
]
`;

exports[`should mount one component correctly 1`] = `
<div>
  Foo Bar
</div>
`;
